File: experiment-runner/examples/greenlab-profiling/codes/gpt/revcomp.py
#!/usr/bin/python3

import sys

def reverse_complement(sequence):
    # Define the complement mapping
    complement = {
        'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 
        'M': 'K', 'R': 'Y', 'W': 'W', 'S': 'S', 
        'Y': 'R', 'K': 'M', 'V': 'B', 'H': 'D', 
        'D': 'H', 'B': 'V', 'N': 'N',
        'a': 'T', 'c': 'G', 'g': 'C', 't': 'A', 
        'm': 'K', 'r': 'Y', 'w': 'W', 's': 'S', 
        'y': 'R', 'k': 'M', 'v': 'B', 'h': 'D', 
        'd': 'H', 'b': 'V', 'n': 'N'
    }
    
    # Use a generator expression to get the complement and then join the results into a string
    # Also, reverse the string using [::-1]
    return ''.join(complement[base] for base in sequence[::-1])

def main():
    # Initialize an empty list to store sequences
    sequences = []
    
    # Iterate over each line in stdin
    for line in sys.stdin:
        # If the line starts with '>', it's a descriptor
        if line.startswith('>'):
            # If we already have sequences, yield them, then clear the list
            if sequences:
                yield ''.join(sequences)
                sequences.clear()
            yield line.strip()
        else:
            # Otherwise, it's a sequence line, so append it to our list
            sequences.append(line.strip())

    # Yield the last sequence, if there is one
    if sequences:
        yield ''.join(sequences)

if __name__ == "__main__":
    # For each sequence, print its reverse complement
    for line in main():
        if line.startswith('>'):
            print(line)
        else:
            print(reverse_complement(line))

File: experiment-runner/examples/greenlab-profiling/codes/gpt/revcomp.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>

std::string reverse_complement(const std::string& sequence) {
    static std::map<char, char> complement = {
        {'A', 'T'}, {'C', 'G'}, {'G', 'C'}, {'T', 'A'},
        {'M', 'K'}, {'R', 'Y'}, {'W', 'W'}, {'S', 'S'},
        {'Y', 'R'}, {'K', 'M'}, {'V', 'B'}, {'H', 'D'},
        {'D', 'H'}, {'B', 'V'}, {'N', 'N'},
        {'a', 'T'}, {'c', 'G'}, {'g', 'C'}, {'t', 'A'},
        {'m', 'K'}, {'r', 'Y'}, {'w', 'W'}, {'s', 'S'},
        {'y', 'R'}, {'k', 'M'}, {'v', 'B'}, {'h', 'D'},
        {'d', 'H'}, {'b', 'V'}, {'n', 'N'}
    };

    std::string result;
    for (auto it = sequence.rbegin(); it != sequence.rend(); ++it) {
        result.push_back(complement[*it]);
    }

    return result;
}

int main() {
    std::string line;
    std::vector<std::string> sequences;

    while (std::getline(std::cin, line)) {
        if (line[0] == '>') {
            if (!sequences.empty()) {
                std::cout << line << std::endl;
                sequences.clear();
            } else {
                std::cout << line << std::endl;
            }
        } else {
            sequences.push_back(line);
        }
    }

    for (const auto& seq : sequences) {
        std::cout << reverse_complement(seq) << std::endl;
    }

    return 0;
}
